challenges-31-plus.yml

```diff
@@ -0,0 +1,85 @@
+    
+    How can we get this information form a controller?
+  explanation: |
+    Whenever Turbo sends a request from one of its Frames, it sends the header
+    `Turbo-Frame:` and the id of the frame it's sending the request from. This
+    is useful for us to know what to do with it!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          ```php
+          $frameId = $request->headers->get('Turbo-Frame', '_top');
+          ```
+        b: |
+          ```php
+          $frameId = $request->headers->get('Turbo-Frame-Id', '_top');
+          ```
+        c: |
+          ```php
+          $frameId = $request->headers->get('Frame', '_top');
+          ```
+        d: |
+          ```php
+          $frameId = $request->headers->get('Turbo', '_top');
+          ```
+      correct: a
+
+modal_frame_close:
+  question: |
+    In this video we've learned about the event `turbo:submit-end`. What
+    is the difference between it and `turbo:submit-start`
+  explanation: |
+    The `turbo:submit-end` event happens *after* a form has been submitted and
+    therefore it contains information about the response from the controller.
+    
+    > Turbo form events are different from other events in that they are fired
+    > from the `<form>` element instead of the `Document`, so you can attach
+    > event listeners to the forms themselves or any of their parent elements.
+  features:
+    choices:
+      a: |
+        `turbo:submit-end` happens after we submit a form and before we
+        get any response back.
+      b: |
+        `turbo:submit-end` happens after we submit a form and contains
+        the HTML of the response.
+      c: |
+        `turbo:submit-end` happens after we submit a form and contains
+        information about the response, like the `success` key.
+      d: |
+        `turbo:submit-end` happens after `turbo:submit-start` and allows
+        us to make asynchronous operations while waiting for a response.
+    correct: c
+
+prevent_frame_rendering:
+  question: |
+    What is the difference between the `turbo:before-fetch-response` and
+    `turbo:submit-end` events?
+  explanation: |
+    The `turbo:before-fetch-request` and `turbo:before-fetch-response` events
+    allow us to cancel the submit of the form, while `turbo:submit-start` and
+    `turbo:submit-end` don't!
+  features:
+    choices:
+      a: |
+        The `turbo:before-fetch-request` event is fired before the form is submitted
+        while `turbo:submit-end` is fired after.
+      b: |
+        The `turbo:before-fetch-request` event is fired before the form is submitted
+        and it allows us to cancel the event while `turbo:submit-end` doesn't.
+      c: |
+        The `turbo:before-fetch-request` event is fired before the form is submitted
+        and it doesn't contain any extra information (such as `success`) while
+        `turbo:submit-end` does contain it.
+      d: |
+        Options b and c are correct.
+    correct: b
```

Comment:
These answers were tricky! They all looked correct to me! 

> a) ... while `turbo:submit-end` is fired after (the form is submitted)

Is this the part that's wrong? Isn't that true?

> c) and it doesn't contain any extra information (such as `success`) while
> `turbo:submit-end` does contain it

Which part of this is wrong? That the `before-fetch-request` doesn't have any extra info? Or that `turbo:submit-end` does contain info?